File: emccd_detect/util/auto_arrange_figures.py
# -*- coding: utf-8 -*-
import warnings

import matplotlib
import matplotlib.pyplot as plt
import numpy as np
from PyQt5 import QtWidgets

matplotlib.rcParams.update({'figure.max_open_warning': 0})


def auto_arrange_figures(nh=0, nw=0, monitor_id=0, rows_first=True):
    """Arrange all figures currently open.

    Call with no arguments to arrange automatically on primary monitor.

    Parameters
    ----------
    nh : int, optional
        Number of images in vertical direction. Defaults to 0.
    nw : int, optional
        Number of images in horizontal direction. Defaults to 0.
    monitor_id : int, optional
        Monitor number, where 0 is primary monitor. Defaults to 0.
    rows_first : bool, optional
        Option to organize from left to right rather than top to bottom.
        Defaults to True.

    Notes
    -----
    In automatic mode, if number of figures is greater than 40, figures will be
    stacked. In manual mode, if grid size is smaller than required for
    accommodating all figures, the function will switch to automatic mode.

    References
    ----------
    Adapted from Lee Jae Jun's Matlab function autoArrangeFigures [1]_.

    .. [1] leejaejun, Koreatech, Korea Republic, 2014.12.13, jaejun0201@gmail.com

    S Miller - UAH - 17-Dec-2018
    """
    task_bar_offset = [30, 50]

    if nh * nw == 0:
        auto_arange = True
    else:
        auto_arange = False

    fignums = plt.get_fignums()
    figs = np.array(list(map(plt.figure, plt.get_fignums())))

    fig_handle = _sort_figure_handles(fignums, figs)
    n_fig = fig_handle.shape[0]
    if n_fig <= 0:
        warnings.warn('Figures are not found.')
        return

    widget = QtWidgets.QDesktopWidget()

    screen = widget.screen(monitor_id)
    screen_sz = screen.geometry()
    scn_w = screen_sz.width() - task_bar_offset[0]
    scn_h = screen_sz.height() - task_bar_offset[1]
    scn_w_begin = screen_sz.left() + task_bar_offset[0]
    scn_h_begin = screen_sz.top() + task_bar_offset[1]

    if auto_arange:
        grid = np.array([[2., 2., 2., 2., 2., 3., 3., 3., 3., 3., 3., 4., 4.,
                          4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4., 4.,
                          4., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5., 5.],
                         [3., 3., 3., 3., 3., 3., 3., 3., 4., 4., 4., 5., 5.,
                          5., 5., 5., 5., 5., 5., 6., 6., 6., 7., 7., 7., 7.,
                          7., 7., 7., 7., 7., 7., 7., 7., 8., 8., 8., 8., 8.]])
        grid = grid.T

        if n_fig <= len(grid):
            nh = grid[n_fig - 1, 0]
            nw = grid[n_fig - 1, 1]
        # If more images than grid length, use largest grid and stack images.
        else:
            nh = grid[-1, 0]
            nw = grid[-1, 1]

        # If sceen is taller than it is wide, switch dimensions of image grid.
        if scn_h > scn_w:
            nh, nw = nw, nh

    fig_width = scn_w / nw
    fig_height = scn_h / nh

    for ipic in range(0, n_fig):
        if rows_first:
            [row, col] = _locate_r(ipic, nh, nw)
        else:
            [row, col] = _locate_c(ipic, nh, nw)

        window = fig_handle[ipic].canvas.manager.window
        window.setGeometry(scn_w_begin + fig_width*(col),
                           scn_h_begin + fig_height*(row),
                           fig_width,
                           fig_height)


def _locate_r(ipic, tot_rows, tot_cols):
    row = (int(np.floor((ipic)/tot_cols)) % tot_rows)
    col = (ipic % tot_cols)

    return [row, col]


def _locate_c(ipic, tot_rows, tot_cols):
    row = (ipic % tot_rows)
    col = (int(np.floor((ipic)/tot_rows)) % tot_cols)

    return [row, col]


def _sort_figure_handles(fignums, figs):
    idx = np.argsort(fignums)
    fig_sorted = figs[idx]

    return fig_sorted

File: emccd_detect_matlab/autoArrangeFigures.m
function autoArrangeFigures(NH, NW, monitor_id, rowsFirst)
% INPUT  :
%        NH : number of grid of vertical direction
%        NW : number of grid of horizontal direction
% OUTPUT :
%
% get every figures that are opened now and arrange them.
%
% autoArrangeFigures selects automatically Monitor1.
% If you are dual(or more than that) monitor user, I recommend to set wide
% monitor as Monitor1.
%
% if you want arrange automatically, type 'autoArrangeFigures(0,0)' or 'autoArrangeFigures()'.
%    But maximum number of figures for automatic mode is 27.
%
% if you want specify grid for figures, give numbers for parameters.
%    but if your grid size is smaller than required one for accommodating
%    all figures, this function changes to automatic mode and if more
%    figures are opend than maximum number, then it gives error.
%
% Notes
%  + 2017.1.20 use monitor id(Adam Danz's idea)
%
% leejaejun, Koreatech, Korea Republic, 2014.12.13
% jaejun0201@gmail.com

if nargin < 4
    rowsFirst = 1;
end

if nargin < 3
    monitor_id = 1;
end

if nargin < 2
    NH = 0;
    NW = 0;
end

task_bar_offset = [30 50];

%%
N_FIG = NH * NW;
if N_FIG == 0
    autoArrange = 1;
else
    autoArrange = 0;
end
figHandle = sortFigureHandles(findobj('Type','figure'));
n_fig = size(figHandle,1);
if n_fig <= 0
    warning('figures are not found');
    return
end

screen_sz = get(0,'MonitorPositions');
screen_sz = screen_sz(monitor_id, :);
scn_w = screen_sz(3) - task_bar_offset(1);
scn_h = screen_sz(4) - task_bar_offset(2);
scn_w_begin = screen_sz(1) + task_bar_offset(1);
scn_h_begin = screen_sz(2) + task_bar_offset(2);

if autoArrange==0
    nh = NH;
    nw = NW;
end

if autoArrange == 1
    grid = [2 2 2 2 2 3 3 3 3 3 3 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4 4;
        3 3 3 3 3 3 3 3 4 4 4 5 5 5 5 5 5 5 5 6 6 6 7 7 7 7 7]';
    
    if n_fig > length(grid)
        warning('too many figures(maximum = %d)',length(grid))
        return
    end
    
    if scn_w > scn_h
        nh = grid(n_fig,1);
        nw = grid(n_fig,2);
    else
        nh = grid(n_fig,2);
        nw = grid(n_fig,1);
    end
end

fig_width = scn_w/nw;
fig_height = scn_h/nh;

for iPic = 1:n_fig
    if rowsFirst
        [row,col] = locateR(iPic,nh,nw);
    else
        [row,col] = locateC(iPic,nh,nw);
    end
    fig_pos = [scn_w_begin + fig_width*(col-1), ...
        scn_h_begin + scn_h-fig_height*row, ...
        fig_width, fig_height];
    set(figHandle(iPic),'OuterPosition',fig_pos);
end
end

function [row,col] = locateR(iPic,totRows,totCols)
row = mod(floor((iPic-1)/totCols),totRows) + 1;
col = mod(iPic-1,totCols) + 1;
end

function [row,col] = locateC(iPic,totRows,totCols)
row = mod(iPic-1,totRows) + 1;
col = mod(floor((iPic-1)/totRows),totCols) + 1;
end

function figSorted = sortFigureHandles(figs)
[tmp, idx] = sort([figs.Number]);
figSorted = figs(idx);
end